{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "948104c4-3cf9-4233-bd62-9ea10f304f94",
   "metadata": {},
   "source": [
    "# BENG183\n",
    "\n",
    "# Acknowledgements: \n",
    "This homework and included tutorials is based on https://github.com/Irenexzwen. Big thanks to her! Check out her website http://wenxingzhao.com/\n",
    "\n",
    "## Homework notes:\n",
    "1) [Working environment \n",
    "setup](https://github.com/Irenexzwen/BIOE183/blob/master/Tutorial1_Preparation.md)\n",
    "2) [Raw Data QC and \n",
    "Cleaning](https://github.com/Irenexzwen/BIOE183/blob/master/Tutorial2_RawData.md)\n",
    "3) **Needs changing** [Mapping and \n",
    "quantification](https://github.com/Irenexzwen/BIOE183/blob/master/Tutorial3_Mapping_and_qualification.md)\n",
    "4) [Differential \n",
    "analysis](https://github.com/Irenexzwen/BIOE183/blob/master/Tutorial4_DE.md)\n",
    "\n",
    "## Raw Sample files:\n",
    "\n",
    "\n",
    "Download the fastq files of the following samples below.\n",
    "\n",
    "### Sample for the homework\n",
    "We will use RNAseq data from [FlyAtlas2 \n",
    "database](http://flyatlas.gla.ac.uk/FlyAtlas2/index.html), which collects \n",
    "hundreds of RNAseq data of drosophila melanogaster. You can search by \n",
    "gene, category or tissue. Here we downloaded 4 samples (female_head x 2, \n",
    "female_midgut x 2).\n",
    "\n",
    "- \n",
    "[[female_head1_R1_raw]](https://drive.google.com/file/d/1T6QDIcvY5qC_E488IjPZEkhvlukgXQbe/view?usp=sharing),[[female_head1_R2_raw]](https://drive.google.com/file/d/1RV6YPV9_KAUr9h-rEdOVRJHPuCV5ozT_/view?usp=sharing)\n",
    "\n",
    "- \n",
    "[[female_midgut1_R1_raw]](https://drive.google.com/file/d/124XZXIhu8Rm8CsvNYx87LVVgDl6UWOt3/view?usp=sharing),[[female_midgut1_R2_raw]](https://drive.google.com/file/d/11rBcLQ-HrYlV3oGgRklkD8-yAX3cCLqY/view?usp=sharing)\n",
    "\n",
    "\n",
    "\n",
    "## Clean Data \n",
    "\n",
    "Follow tutorail 2 to clean reads in fastq files.\n",
    "\n",
    "## Reference genome.fa / transcriptome.fa / gtf\n",
    "We usually download the reference data from \n",
    "[ensemble](https://uswest.ensembl.org/info/data/ftp/index.html). You \n",
    "search \"drosophila\" and choose DNA / cDNA / gtf, then you use a `wget` to \n",
    "download.\n",
    "- drosophila genome: \n",
    "ftp://ftp.ensembl.org/pub/release-97/fasta/drosophila_melanogaster/dna/Drosophila_melanogaster.BDGP6.22.dna.toplevel.fa.gz\n",
    "\n",
    "- drosophila transcriptome: \n",
    "ftp://ftp.ensembl.org/pub/release-97/fasta/drosophila_melanogaster/cdna/Drosophila_melanogaster.BDGP6.22.cdna.all.fa.gz\n",
    "\n",
    "- drosophila gtf: \n",
    "ftp://ftp.ensembl.org/pub/release-97/gtf/drosophila_melanogaster/Drosophila_melanogaster.BDGP6.22.97.chr.gtf.gz\n",
    "\n",
    "## Index files\n",
    "Pre-computed index files: [download \n",
    "here](https://drive.google.com/open?id=1CT-iZ2PzPwWa44KxhjYsv4pD5GnnJStg)\n",
    "\n",
    "## Mapping .bam file\n",
    "\n",
    "Follow tutorial 3 to produce bam files.\n",
    "\n",
    "## FeatureCounts Count table\n",
    "\n",
    "Follow tutorial 3 to produce feature counts table.\n",
    "\n",
    "You can download feature counts for all samples [here]('https://drive.google.com/file/d/1E_QnRJ_b98H_hINhukBXdqxd0Q2Qcqu4/view?usp=sharing')\n",
    "\n",
    "## Resources:\n",
    "#### Tutorials:\n",
    "[1] Weill Cornell Medical Colledge: \n",
    "http://chagall.med.cornell.edu/RNASEQcourse/\n",
    "\n",
    "#### Software manuals:\n",
    "- [Bioconda](https://www.youtube.com/watch?v=lGa9PCSH5IU) starting from \n",
    "3:30. \n",
    "- [fastQC manual](https://dnacore.missouri.edu/PDF/FastQC_Manual.pdf)\n",
    "- [fastp manual](https://github.com/OpenGene/fastp)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
